src/vs/workbench/parts/terminal/electron-browser/terminalInstance.ts

```diff
@@ -676,32 +677,43 @@ export class TerminalInstance implements ITerminalInstance {
 		}
 	}
 
-	public preparePathForTerminalAsync(path: string): Promise<string> {
+	public preparePathForTerminalAsync(originalPath: string): Promise<string> {
 		return new Promise<string>(c => {
-			const hasSpace = path.indexOf(' ') !== -1;
+			const exe = this.shellLaunchConfig.executable;
+			if (!exe) {
+				c(originalPath);
+				return;
+			}
+
+			const hasSpace = originalPath.indexOf(' ') !== -1;
+
+			const pathBasename = path.basename(exe, '.exe');
+			const isPowerShell = pathBasename === 'pwsh' ||
+				this.title === 'pwsh' ||
+				pathBasename === 'powershell' ||
+				this.title === 'powershell';
+
+			if (hasSpace && isPowerShell) {
+				c(`& '${originalPath.replace('\'', '\'\'')}'`);
```

Comment:
Does PowerShell just not let you run paths that have single quotes in them?

```
touch foo\'
```

Reply:
Not sure I understand.

In this case would `'` be the name of the file?

to do that in PowerShell, you'd either need to do:

```
touch "foo\'"
# or
touch 'foo\'''
```

The second looks weird... basically in PowerShell, strings that use `' '` do not give you the ability to escape characters... Strings with `" "` DO allow you to escape characters (think `\n` only in PowerShell instead of `\` we use ` backtick so `n, `t, etc)

However... that means that you couldn't escape `'` in a `' '` string... so PowerShell has this concept of "escaping `'` in `' '` strings by adding ANOTHER `'`. Hence... 3 `'` in the second example:
1. to escape the middle one
2. the actual `'` we wanna insert
3. the end of the `' '` string

Does that make sense?

Reply:
`touch foo\'` on Linux will create a file called `foo'`, I suspect this edge case may not work with your change:

![image](https://user-images.githubusercontent.com/2193314/50797506-e4cd7380-1289-11e9-8913-571729bad0ea.png)

Unless this is fine: `& 'foo'''`

Reply:
Yes `& 'foo'''` would work.

Here I try to run a `foo'.ps1`:

```
# throw a log statement in a foo'.ps1 - note the backtick is the escape character for PowerShell
> 'Write-Host "asdf"' > foo`'.ps1

# foo'.ps1 is in my directory
> ls
LICENSE					ThirdPartyNotices.txt			build.ps1				package					test
NuGet.config				appveyor.yml				examples				protobuf				tools
README.md				azure-functions-powershell-worker.sln	foo'.ps1				src

# execute the foo'.ps1
> & './foo''.ps1'
asdf

Reply:
Oh! Wow I finally understand what you're saying...

If we have a path that's like:

`/Users/tyriar/foo'.ps1` it's going to fail because it's not going to actually escape it because there's no space!